Propagate Smallrye Context when switching REST Client context

Sgitario opened the pull request:
The problem is that Opentelemetry creates the span context in MDCEnabledContextStorage when there is no VertxContext yet. 

Then, the span context of MDCEnabledContextStorage is propagated to the VertxContext in every operator of an Uni/Multi ... action. Before the changes in https://github.com/quarkusio/quarkus/pull/32852, this was done in https://github.com/quarkusio/quarkus/pull/32852/commits/272532591b2bd437ed91dbb8e73d6336d65ea781#diff-03d3c6adcdcc01b3a1f8daccc773adc6ee1c598afa245a445a24383a77f3b3a4L92.

After the changes of https://github.com/quarkusio/quarkus/pull/32852, we're switching the context but without using any Uni/Multi/... instance, so the span context is not attached to the one that will be created later in the VertxContext, and this is why we got two spans. 

Personally, I don't like that the spancontext is being propagated only using the Uni/Multi/... instances (or at least, I could not find any other way to do it). I think we should implement a listener interface when switching of contexts (tho MDCEnabledContextStorage is an opentelemetry thing).

To mimic what Smallrye does internally when calling the Uni/Multi operators, I directly call the `Infrastructure.decorate` method which fixes the issue. 

Fix https://github.com/quarkusio/quarkus/issues/34212 Relates https://github.com/quarkusio/quarkus/pull/32852

Sgitario commented:
Btw, I didn't add any tests, but you can use the reproducer to double check it works. 
I'm planning to add tests tomorrow morning.

geoand commented:
LGTM, but I'll refrain from approving until we have a test that previously failed and now doesn't 😎

brunobat commented:
These split spans happen only on startup, if a rest client call is performed, right?

Sgitario committed:
Propagate Smallrye Context when switching REST Client context

The problem is that Opentelemetry creates the span context in MDCEnabledContextStorage when there is no VertxContext yet. 

Then, the span context of MDCEnabledContextStorage is propagated to the VertxContext in every operator of an Uni/Multi ... action. Before the changes in https://github.com/quarkusio/quarkus/pull/32852, this was done in https://github.com/quarkusio/quarkus/pull/32852/commits/272532591b2bd437ed91dbb8e73d6336d65ea781#diff-03d3c6adcdcc01b3a1f8daccc773adc6ee1c598afa245a445a24383a77f3b3a4L92.

After the changes of https://github.com/quarkusio/quarkus/pull/32852, we're switching the context but without using any Uni/Multi/... instance, so the span context is not attached to the one that will be created later in the VertxContext, and this is why we got two spans. 

Personally, I really dislike that the spancontext is being propagated only using the Uni/Multi/... instances (or at least, I could not find any other way to do it). I think we should implement a listener interface when switching of contexts (tho MDCEnabledContextStorage is an opentelemetry thing).

To mimic what Smallrye does internally when calling the Uni/Multi operators, I directly call the Infrastructure.decorate method which fixes the issue. 

Fix https://github.com/quarkusio/quarkus/issues/34212
Relates https://github.com/quarkusio/quarkus/pull/32852

Sgitario commented:
> LGTM, but I'll refrain from approving until we have a test that previously failed and now doesn't sunglasses

PR updated with one test that verifies the fix. 

> These split spans happen only on startup, if a rest client call is performed, right?

Yes, this was only happening when there is no VertxContext started (for ex: at startup).

brunobat commented:
@geoand, It was this issue: https://github.com/quarkusio/quarkus/issues/32923
They are related with startup and the delayed instantiation of the OTel SDK, but the root cause is not the same, I believe.

Sgitario commented:
@geoand @brunobat are you ok with merging this or do you want something else to check beforehand?

geoand commented:
+1 for merging

gsmet commented:
@Sgitario @geoand should we backport this? I feel like we do but maybe I'm missing something?

Sgitario commented:
> We definitely should!

+100
Sorry, I forgot adding the label.

gsmet commented:
Thanks!
